### PR TITLE
Reimplementing destructuring

### PR DESCRIPTION
**Pull request by yloiseau**
Trying to reimplement the destructuring feature better (see #524 for a full introduction to issues of current implementation).

### CODE REVIEW

File: TODO.md [hunk 1]
```diff
@@ -21,6 +21,8 @@ Trying to reimplement the destructuring feature better (see #524 for a full intr
     - [x] lazy lists
     - [x] ranges
     - [x] When IR
+    - [x] Binary operation IR
```
Comment (jponge) [inline TODO.md:24]: Make sure to remove this file 😉

File: src/main/java/gololang/HeadTail.java [hunk 2]
```diff
@@ -76,4 +78,38 @@
...
+   * New style destructuring helper.
+   *
+   * New style destructuring must be exact. The number of variables to be affected is thus checked against the number of
+   * members of the structure.
+   *
+   * @param number number of variable that will be affected.
+   * @param substruct whether the destructuring is complete or should contains a sub structure.
+   * @return an array containing the values to assign.
+   */
+  default Object[] __$$_destruct(int number, boolean substruct, Object[] toSkip) {
```
Comment (jponge) [inline src/main/java/gololang/HeadTail.java:92]: Lacks a `@param` entry for `toSkip`

File: src/main/java/gololang/Range.java [hunk 3]
```diff
@@ -81,4 +85,54 @@
...
+   * New style destructuring helper.
+   *
+   * New style destructuring must be exact. The number of variables to be affected is thus checked against the number of
+   * members of the structure.
+   *
+   * @param number number of variable that will be affected.
+   * @param substruct whether the destructuring is complete or should contains a sub structure.
+   * @return an array containing the values to assign.
+   */
+  default Object[] __$$_destruct(int number, boolean substruct, Object[] toSkip) {
```
Comment (jponge) [inline src/main/java/gololang/Range.java:99]: Lacks a `@param` entry for `toSkip`

**Review by jponge — CHANGES_REQUESTED** [review]

File: TODO.md (hunk 1 above)
Comment (yloiseau) [reply TODO.md:24]: Yes, internal reminder... I'll remove and squash

File: src/main/java/gololang/HeadTail.java (hunk 2 above)
Comment (yloiseau) [reply src/main/java/gololang/HeadTail.java:92]: :+1:

File: src/main/java/gololang/Range.java (hunk 3 above)
Comment (yloiseau) [reply src/main/java/gololang/Range.java:100]: :+1:

**Review by jponge — APPROVED** [review]

**Pull request merged**